scanpy/experimental/pp/_highly_variable_genes.py

```diff
@@ -0,0 +1,341 @@
+        adata_subset = adata_subset[:, nonzero_genes]
+
+        if layer is not None:
+            X_batch = adata_subset.layers[layer]
+        else:
+            X_batch = adata_subset.X
+
+        # Prepare clipping
+        if clip is None:
+            n = X_batch.shape[0]
+            clip = np.sqrt(n)
+        if clip < 0:
+            raise ValueError("Pearson residuals require `clip>=0` or `clip=None`.")
+
+        if sp_sparse.issparse(X_batch):
+            sums_genes = np.sum(X_batch, axis=0)
+            sums_cells = np.sum(X_batch, axis=1)
+            sum_total = np.sum(sums_genes).squeeze()
+        else:
+            sums_genes = np.sum(X_batch, axis=0, keepdims=True)
+            sums_cells = np.sum(X_batch, axis=1, keepdims=True)
+            sum_total = np.sum(sums_genes)
+
+        # Compute pearson residuals in chunks
+        residual_gene_var = np.empty((X_batch.shape[1]))
+        for start in np.arange(0, X_batch.shape[1], chunksize):
+            stop = start + chunksize
+            mu = np.array(sums_cells @ sums_genes[:, start:stop] / sum_total)
+            X_dense = X_batch[:, start:stop].toarray()
+            residuals = (X_dense - mu) / np.sqrt(mu + mu ** 2 / theta)
+            residuals = np.clip(residuals, a_min=-clip, a_max=clip)
+            residual_gene_var[start:stop] = np.var(residuals, axis=0)
+
+        # Add 0 values for genes that were filtered out
+        zero_gene_var = np.zeros(np.sum(~nonzero_genes))
+        residual_gene_var = np.concatenate((residual_gene_var, zero_gene_var))
+        # Order as before filtering
+        idxs = np.concatenate((np.where(nonzero_genes)[0], np.where(~nonzero_genes)[0]))
+        residual_gene_var = residual_gene_var[np.argsort(idxs)]
+        residual_gene_vars.append(residual_gene_var.reshape(1, -1))
+
+    residual_gene_vars = np.concatenate(residual_gene_vars, axis=0)
+
+    # Get cutoffs and define hvgs per batch
+    residual_gene_vars_sorted = np.sort(residual_gene_vars, axis=1)
+    cutoffs_per_batch = residual_gene_vars_sorted[:, -n_top_genes]
+    highly_variable_per_batch = np.greater_equal(
+        residual_gene_vars.T, cutoffs_per_batch
+    ).T
+
+    # Merge hvgs across batches
+    highly_variable_nbatches = np.sum(highly_variable_per_batch, axis=0)
+    highly_variable_intersection = highly_variable_nbatches == n_batches
+
+    # Get rank per gene within each batch
+    # argsort twice gives ranks, small rank means most variable
+    ranks_residual_var = np.argsort(np.argsort(-residual_gene_vars, axis=1), axis=1)
+    ranks_residual_var = ranks_residual_var.astype(np.float32)
+    ranks_residual_var[ranks_residual_var >= n_top_genes] = np.nan
+    ranks_masked_array = np.ma.masked_invalid(ranks_residual_var)
+    # Median rank across batches,
+    # ignoring batches in which gene was not selected
+    medianrank_residual_var = np.ma.median(ranks_masked_array, axis=0).filled(np.nan)
```

Comment:
I have taken the logic mostly from the seuratV3 function:

https://github.com/theislab/scanpy/blob/63b42e4bff46a9e1386abfede4adeef3a8100c7b/scanpy/preprocessing/_highly_variable_genes.py#L119-L130

I computed some things in a few more lines before, now adapted to the SeuratV3 way, hoping to make this more clear.

Similarly, I use median rank because SeuratV3 also uses median rank here.